### PR TITLE
Publish docs using Sphinx.

### LINKED ISSUE #5: Docs to RTD

### PR DESCRIPTION
**Pull request by jamadden**
Some doc formatting fixes and cleanups. Add quite a bit of docs about namespaces and the various ones we supply.

Fixes #5 

Docs can be viewed at http://zopetraversing.readthedocs.io/en/rtd-docs/

### CODE REVIEW

File: src/zope/traversing/api.py [hunk 1]
```diff
@@ -26,22 +29,8 @@
...
-
-    A segment may not start with a '/' because that would be confused
-    with an absolute path. A segment may not end with a '/' because we
-    do not allow '/' at the end of relative paths.  A segment may
-    consist of . or .. to mean "the same place", or "the parent path"
-    respectively. A '.' should be removed and a '..' should cause the
-    segment to the left to be removed.  joinPath('/', '..') should
-    raise an exception.
+    """
+    Join the given relative paths to the given path.
```
Comment (mgedmin) [inline src/zope/traversing/api.py]: ~~Did you mean to delete all this text?~~

This seemed peculiar before I saw the interface `__doc__` copying at the end of the module.

File: src/zope/traversing/api.py [hunk 2]
```diff
@@ -58,32 +47,21 @@ def joinPath(path, *args):
...
-
-    Raises LocationError if path cannot be found
-
-    Note: calling traverse with a path argument taken from an untrusted
-          source, such as an HTTP request form variable, is a bad idea.
-          It could allow a maliciously constructed request to call
-          code unexpectedly.
-          Consider using traverseName instead.
+    """
     """
```
Comment (mgedmin) [inline src/zope/traversing/api.py]: ~~Was this change intentional?  It doesn't look intentional.~~

This seemed peculiar before I saw the interface `__doc__` copying at the end of the module.

I still don't think having an explicit empty docstring was intentional.

File: src/zope/traversing/api.py [hunk 3]
```diff
@@ -92,21 +70,8 @@ def traverse(object, path, default=_marker, request=None):
...
-    argument. If you do not, the given object will be adapted to ITraversable.
-
-    'request' is passed in when traversing from presentation code. This
-    allows paths like @@foo to work.
-
-    Raises LocationError if path cannot be found and 'default' was
-    not provided.
-
+    """
+    Traverse a single step 'name' relative to the given object.
```
Comment (mgedmin) [inline src/zope/traversing/api.py]: ~~It looks like useful description is being lost.~~

This seemed peculiar before I saw the interface `__doc__` copying at the end of the module.

File: src/zope/traversing/api.py [hunk 4]
```diff
@@ -122,17 +87,15 @@ def traverseName(obj, name, default=_marker, traversable=None, request=None):
 
 
 def getName(obj):
-    """Get the name an object was traversed via
+    """
+    Get the name an object was traversed via
```
Comment (mgedmin) [inline src/zope/traversing/api.py]: Let's add a period at the end of the sentence while we're cleaning things up?

File: src/zope/traversing/api.py [hunk 5]
```diff
@@ -122,17 +87,15 @@ def traverseName(obj, name, default=_marker, traversable=None, request=None):
...
 
 
 def getParent(obj):
-    """Returns the container the object was traversed via.
-
-    Returns None if the object is a containment root.
-    Raises TypeError if the object doesn't have enough context to get the
-    parent.
+    """
+    Returns the container the object was traversed via.
```
Comment (mgedmin) [inline src/zope/traversing/api.py]: ~~Information about containment root and raised exceptions seems like it could be useful?~~

This seemed peculiar before I saw the interface `__doc__` copying at the end of the module.

File: src/zope/traversing/api.py [hunk 6]
```diff
@@ -157,11 +120,9 @@ def getParent(obj):
 
 
 def getParents(obj):
-    """Returns a list starting with the given object's parent followed by
+    """
+    Returns a list starting with the given object's parent followed by
     each of its parents.
-
-    Raises a TypeError if the context doesn't go all the way down to
-    a containment root.
```
Comment (mgedmin) [inline src/zope/traversing/api.py]: ~~That seems like useful information to me.~~

This seemed peculiar before I saw the interface `__doc__` copying at the end of the module.

File: src/zope/traversing/api.py [hunk 7]
```diff
@@ -223,3 +182,9 @@ def canonicalPath(path_or_object):
 
 # import this down here to avoid circular imports
 from zope.traversing.adapters import traversePathElement
+
+# Synchronize the documentation.
+for name in ITraversalAPI.names():
+    if name in globals():
+        globals()[name].__doc__ = ITraversalAPI[name].__doc__
+del name
```
Comment (mgedmin) [inline src/zope/traversing/api.py:211]: Oh, I see!

I like the reduction in duplication, but maybe add "See ITraversalAPI for more details" to each of the docstrings that are replaced in this fashion?  Some people (e.g. me) will look at the source code rather than read the docs on a web page, and the loss of detail should be offset by a hint where to find more detail.

File: src/zope/traversing/browser/interfaces.py [hunk 8]
```diff
@@ -39,6 +45,16 @@ def breadcrumbs():
 
 
 class IAbsoluteURLAPI(Interface):
+    """
+    The api to compute absolute URLs of objects.
```
Comment (mgedmin) [inline src/zope/traversing/browser/interfaces.py]: API should be capitalized.

File: src/zope/traversing/namespace.py [hunk 9]
```diff
@@ -176,7 +221,7 @@ def queryResource(site, name, request, default=None):
     # resource to do this.  We still return the proxied resource.
     r = removeSecurityProxy(resource)
 
-    r.__parent__ = site
+    r.__parent__ = context
```
Comment (mgedmin) [inline src/zope/traversing/namespace.py:229]: I'm not convinced by this renaming.  As far as I remember, in Zope 3 resources are only directly available on sites, not on arbitrary traversal contexts.

File: src/zope/traversing/namespace.py [hunk 10]
```diff
@@ -11,7 +11,48 @@
 # FOR A PARTICULAR PURPOSE.
 #
 ##############################################################################
-"""URL Namespace Implementations
+"""
+URL Namespace Implementations
+
+A URL Namespace is usually a path segment that looks like ``++ns++name``.
+(It can also look like ``@@name``, which is a
+shortcut for ``++view++name``. See :func:`nsParse` for details.)
```
Comment (mgedmin) [inline src/zope/traversing/namespace.py]: My inner pedant wants to say that namespace is *not* a path segment.  I'll try to suggest a rephrasing.  Hmm, maybe

> A namespace can be used in a path segment like ``++ns++name`` to disambiguate (e.g. items versus attributes), or access a different set of traversable names (e.g. views, accessible via ``++view++name``, or its special shortcut spelling ``@@name``).

**Review by mgedmin — APPROVED** [review]
LGTM with minor comments

File: src/zope/traversing/api.py (hunk 2 above)
Comment (jamadden) [reply src/zope/traversing/api.py]: Yikes, no it wasn't, I was trying to leave at least the topic sentence. Good catch.

File: src/zope/traversing/api.py (hunk 4 above)
Comment (jamadden) [reply src/zope/traversing/api.py]: Good idea!

File: src/zope/traversing/api.py (hunk 7 above)
Comment (jamadden) [reply src/zope/traversing/api.py:211]: Good idea! I agree it's important to find a balance.

File: src/zope/traversing/namespace.py (hunk 10 above)
Comment (jamadden) [reply src/zope/traversing/namespace.py]: Good point, that was sloppy wording on my  part. How about:

> URL namespaces are an extensible mechanism to provide additional control over traversal (for example, disambiguating item versus attribute access) or access to an additional set of traversable names (such as :class:`registered views <view>` or :class:`path adapters
<adapter>`).  This mechanism is used for path segments that look like ``++ns++name``.  (It is also used for segments like ``@@name``, which is a shortcut for ``++view++name``.  See :func:`nsParse` for details.)

File: src/zope/traversing/namespace.py (hunk 9 above)
Comment (jamadden) [reply src/zope/traversing/namespace.py:229]: From configure.zcml, that namespace appears to be available for any type of object:
```xml
<adapter
    name="resource"
    for="* zope.publisher.interfaces.IRequest"
    provides="zope.traversing.interfaces.ITraversable"
    factory="zope.traversing.namespace.resource"
    />
```

**Comment by jamadden** [comment]
Thanks for the great comments!

File: src/zope/traversing/namespace.py (hunk 9 above)
Comment (jamadden) [reply src/zope/traversing/namespace.py:229]: One place `++resource` is used turns out to be in [zope.viewlet](http://zopeviewlet.readthedocs.io/en/rtd-docs/viewlet.html) for CSS and JavaScript links. IIUC, this namespace only looks at the request object because it has to do with skins/layers.

Comment (mgedmin) [reply src/zope/traversing/namespace.py:229]: I may be conflating things.  The way I remember things work, you use something like

    <link rel="stylesheet" tal:attributes="href context/++resource++bar.css" />

in a template, and it'll render to a URL like `http://localhost:8080/@@/bar.css`.  The nameless view for actually serving resources is registered only for ISite.

I'm currently baffled by which part of this chain is making sure the `__parent__` of the resource is actually a site... Oh!  There's a custom IAbsoluteURL implementation for resources that uses `getSite()` instead of the resource's own `__parent__` that does this.

Ok, your change is correct and good.  Thank you!
